blueprint/core/blueprint_GenericEditor.cpp

```diff
@@ -25,9 +29,18 @@ namespace blueprint
         for (auto& p : proc.getParameters())
             p->addListener(this);
 
-        // Setup the ReactApplicationRoot callbacks and evaluate the supplied JS code/bundle
-        registerAppRootCallbacks();
+        // Setup the hot reloading callbacks
+        harness.onBeforeAll = [this]() { beforeBundleEvaluated(); };
+        harness.onAfterAll = [this]() { afterBundleEvaluated(); };
+
+        // TODO: Should we just watch a bunch of files and then ask the harness to `start` manuall?
```

Comment:
@JoshMarler I want to carry the initializer_list comment idea in here.

Basically you can see here that, as written, I have to tell the harness to `watch` my bundle and then manually tell appRoot to evaluate. The annoying part there is that at his point, none of the bundles have _changed_ as far as the harness is concerned, so it doesn't run its `onBefore*`/`onAfter*` callbacks, which means I have this manual invocation of `beforeBundleEvaluated`/`afterBundleEvaluated` below. This feels a little clunky.

Here's the change I want to propose: instead of starting the harness for you by default, we ask that you first `watch` all the bundles you want watched, then call `start`. At that time, we run the callback hooks for you in the order expected and kick off that initial call to appRoot.evaluate.

That would mean that the following few lines look as follows:
```
harness.watch(bundleFile);
harness.start(); // Will kick off the first few callback hooks in order and call appRoot.evaluate
```
OR, supposing I have multiple bundle files:
```
harness.watch({bundleFileA, bundleFileB, bundleFileC});
harness.start(); // Will kick off the first few callback hooks in order and call appRoot.evaluate
```

What do you think?

Reply:
Yeah interesting. Think this API looks fine. It does force a rather different flow of events for debug/release builds if that is the distinction that people use to switch hot-reload on and off. So I do wonder if the GenericEditor should switch between the two approaches based on `JUCE_DEBUG`. If only to give and example of a likely setup, i.e. in release builds you still need to manual `beforeBundleEvaluated`, `evaluate`, `afterBundleEvaluated` chain?

Reply:
Hm, that's a great point...

The other option is to have `HotReloadingHarness` just _not_ run it's file watcher timer in release. For example:

```
void HotReloadingHarness::start()
{
    if (onBeforeAll) { onBeforeAll(); }

    for (const auto& f : fileWatcher->getWatchedFiles())
    {
        if (onBeforeEach) { onBeforeEach(f); }
        appRoot.evaluate(f);
        if (onAfterEach) { onAfterEach(f); }
    }

    if (onAfterAll) { onAfterAll(); }

#if JUCE_DEBUG
    if (fileWatcher)
        fileWatcher->start();
#endif
}
```

That way you don't _need_ to do anything as an end user, but again if you do need some particular behavior rewriting a custom HotReloadingHarness should be pretty easy

What do you think?

Reply:
Yeah I did think about that as an option. If we go with that option it should definitely be clearly documented and I wonder if a `HotReloadingHarness` is still the correct name at the point where it serves two purposes to some degree (i.e. saving the user boiler-plate). `AppReloadHarness` ?

Reply:
That's picky. It doesn't need to block.

Reply:
Yea, good point. Maybe just `AppHarness` since in release it won't actually reload anything?

I'm down with that!

Reply:
Hahaha yeah do it.

Reply:
Approved on my side with that addition.